### PR TITLE
.github: add workflow to spot .github/ changes

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
Ensure that we explicitly ack changes to .github/ by way of setting a
label.

### CODE REVIEW

**Comment by marusak** [comment]
OOI what is point of this?
And do I read it correctly that some workflow just fails when you touch .github? (and you cannot make it green)

File: .github/workflows/github-metadata.yml
Comment (martinpitt) [inline .github/workflows/github-metadata.yml:8]: Bikeshedding: "Protect GitHub workflow changes"

File: .github/workflows/github-metadata.yml
Comment (martinpitt) [inline .github/workflows/github-metadata.yml:9]: Maybe "approved-workflow-change"?

File: .github/workflows/github-metadata.yml
Comment (martinpitt) [inline .github/workflows/github-metadata.yml:13]: What does this do?

File: .github/workflows/github-metadata.yml
Comment (martinpitt) [inline .github/workflows/github-metadata.yml:20]: Hmmm.. You have a really strict timeout. Could this perhaps just do something like

    curl --silent --show-error -L https://github.com/cockpit-project/cockpit/pull/15771.diff | grep '^+++ .*/\.github/'

(like for [this PR](https://github.com/cockpit-project/cockpit/pull/15771.diff)) instead of cloning the whole repo just for this?

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! Some bikeshedding. I'm not 100% sure yet whether we  want this or this is just additional bureaucracy -- we review PRs before landing anyway, and PR workflows don't even run for first-time contributors automatically (there is an explicit approval button now). But it's an interesting idea.

**Comment by martinpitt** [comment]
If we want to put any kind of reliance on this, there is an important part missing, though: Every PR update must automatically remove the label again. Otherwise you could send a PR, get it reviewed, we set the label, then you force-push and add something malicious. This could even trick us into *not* carefully reviewing the force-push again.

File: .github/workflows/github-metadata.yml
Comment (allisonkarlitskaya) [reply .github/workflows/github-metadata.yml:8]: There's other things (like codeowners and so) that can live in `.github/`.  It's not just workflows.

File: .github/workflows/github-metadata.yml
Comment (allisonkarlitskaya) [reply .github/workflows/github-metadata.yml:9]: ditto on workflow vs metadata bikeshed.

File: .github/workflows/github-metadata.yml
Comment (allisonkarlitskaya) [reply .github/workflows/github-metadata.yml:13]: This makes it so the `git init` below knows where to write to.  It's basically an alternative to separately passing `--git-dir` to every single git command.

File: .github/workflows/github-metadata.yml
Comment (allisonkarlitskaya) [reply .github/workflows/github-metadata.yml:20]: That seems a bit racy....  I'd prefer to run on the object we've been handed to run on....

**Comment by allisonkarlitskaya** [comment]
> If we want to put any kind of reliance on this, there is an important part missing, though: Every PR update must automatically remove the label again. Otherwise you could send a PR, get it reviewed, we set the label, then you force-push and add something malicious. This could even trick us into _not_ carefully reviewing the force-push again.

This part I absolutely agree with.  I'll add that.

File: .github/workflows/github-metadata.yml
Comment (martinpitt) [reply .github/workflows/github-metadata.yml:8]: OK, fair enough. Calling it "meta"data is still a bit weird though. And our other "name:"s are capitalized.

File: .github/workflows/github-metadata.yml
Comment (martinpitt) [reply .github/workflows/github-metadata.yml:20]: Can you then please at least do shallow fetches of the two things you want to compare? And bump the timeout a bit?

**Comment by martinpitt** [comment]
Let's discuss that in tomorrow's team meeting, I put that on the agenda.

**Comment by allisonkarlitskaya** [comment]
We discussed this in the meeting, and although some people had some reservations about if we *really* need this (since we should catch these kinds of changes in reviews), nobody thought that it would really hurt anything to have it.

Additionally, there is at least one scenario where this is helpful: when you review a large patch in-depth the first time, and then request changes, it's not uncommon to just go to the changed section to make sure that you got what you asked for.  That's maybe not the best practice for dealing with external contributions, but we're all probably guilty of it from time to time.  Additionally, if the submitter rebased their patch in the meanwhile, the github-generated "forcepushed" diff is basically unreadable.  That's a window where someone could sneak in nasty changes, and this workflow would catch that.

Also: the potential cost of someone landing malicious changes to our `.github` directory is incredibly high.

**Comment by allisonkarlitskaya** [comment]
Changed: the label is called `.github-changes` now, which is pretty neutral in terms of terminology squabbling, and the label now gets unset on any code changes.

I've also generalised the name of the workflow so that we could add additional "safety checks" here if we wanted to.

File: .github/workflows/github-metadata.yml
Comment (martinpitt) [reply .github/workflows/github-metadata.yml]: Repeating in new review, as this is labeled as outdated.

File: .github/workflows/reposchutz.yml
Comment (martinpitt) [inline .github/workflows/reposchutz.yml:40]: So the intention here is to not fetch the whole repo (as a naïve `git clone` would do), but only the bits we need. I wonder how this can work, though, as it only fetches one half, and not `github.sha`. Either this only works if the PR is fully rebased on current master (and thus `github.sha` is part of the history of `github.event.pull_request.head.sha`), or somehow the fetch pulls all the heads/branches anyway (in which case there's nothing we can do).

Did you try this with an "outdated" branch?

File: .github/workflows/reposchutz.yml
Comment (martinpitt) [inline .github/workflows/reposchutz.yml:10]: There is a little too much going on here for this timeout -- let's make it 5 mins?

File: .github/workflows/reposchutz.yml
Comment (martinpitt) [inline .github/workflows/reposchutz.yml:34]: I find a label with a dot a bit weird -- can we drop that?

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks!

File: .github/workflows/reposchutz.yml
Comment (allisonkarlitskaya) [reply .github/workflows/reposchutz.yml:40]: The `github.sha` is according to the docs, the base commit of the PR, which is also one of the parent commits of the merge commit that github makes for the PR (as I understand it).

But it also means I fetch the full history, which is not necessary.  Probably it makes sense to replace this with two separate fetches, of the SHAs that we are interested in, and `--depth=1`.

File: .github/workflows/reposchutz.yml
Comment (allisonkarlitskaya) [reply .github/workflows/reposchutz.yml:10]: Sure

File: .github/workflows/reposchutz.yml
Comment (allisonkarlitskaya) [reply .github/workflows/reposchutz.yml:34]: Why?  The github docs even mention that label names could happily contain emoji.  Do you have a suggestion for an alternative?

Comment (martinpitt) [reply .github/workflows/reposchutz.yml:34]: It's not related to what the API allows, this was a purely aesthetical remark. Just "github-changes"? But this is of course not that important, not a strong opinion.

File: .github/workflows/reposchutz.yml
Comment (martinpitt) [reply .github/workflows/reposchutz.yml:40]: Sounds good!

File: .github/workflows/reposchutz.yml
Comment (allisonkarlitskaya) [reply .github/workflows/reposchutz.yml:34]: I'd prefer to leave it as is, in that case.  ".github" is the thing that's changing, so it makes sense to me to name it as such :)

**Review by martinpitt — APPROVED** [review]
reposchutz! @major would be so proud :grin:

**Comment by major** [comment]
@martinpitt

**Pull request merged**
